### PR TITLE
SAM/Deploy: fix recently used bucket setting

### PR DESCRIPTION
**Pull request by JadenSimon**
## Problem
We specified the `manuallySelectedBuckets` as an object but we were storing it as a string. 

## Solution
* Write the setting as an object. A new interface `SavedBuckets` was added to clarify the structure of saved buckets.
* Check for previously saved stringified buckets and parse them. Failed parses will reset the setting.
* Add two new tests to check the above.

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by justinmk3** [comment]
- Instead of changing the `type` in package.json why don't we change our logic that stores it, to actually store an object? 
    - Another reason I would tend to avoid changing the `type` is, it would require a "migration" of old configs. Though we should likely code defensively around this in any case.
- needs a `npm run newChange`

**Comment by JadenSimon** [comment]
Good point. I'll change it to an 'object' and see how it looks in the settings.

**Comment by JadenSimon** [comment]
Prettier changed a lot of random stuff. The key point is `JSON.stringify` and `JSON.parse` were removed.

File: src/lambda/commands/deploySamApplication.ts [hunk 1]
```diff
@@ -116,23 +116,22 @@ export async function deploySamApplication(
...
                 } else {
                     settings.writeSetting(
                         CHOSEN_BUCKET_KEY,
-                        JSON.stringify({
-                            ...buckets,
+                        {
+                            ...bucketsJson,
                             [profile]: {
-                                ...(buckets[profile] ? buckets[profile] : {}),
+                                ...(bucketsJson[profile] ? bucketsJson[profile] : {}),
```
Comment (justinmk3) [inline src/lambda/commands/deploySamApplication.ts:131]: We should probably move this logic into `writeFreshJson()` (and rename it to `save()` or something)

**Comment by justinmk3** [comment]
I wonder why we store this in settings instead of globalStorage.

**Comment by JadenSimon** [comment]
> I wonder why we store this in settings instead of globalStorage.

We could do this later on by creating a new `SettingsConfiguration` class that reads/writes to a json file in our `globalStoragePath`. Would have to add logic to migrate from the old settings to the new one of course.

File: src/lambda/wizards/samDeployWizard.ts [hunk 2]
```diff
@@ -190,6 +194,47 @@ function getSingleResponse(responses: vscode.QuickPickItem[] | undefined): strin
...
 }
 
+/**
+ * The toolkit used to store saved buckets as a stringified JSON object. To ensure compatability,
+ * we need to check for this and convert them into objects.
+ */
+export function readSavedBuckets(settings: SettingsConfiguration): SavedBuckets | undefined {
+    try {
+        const buckets = settings.readSetting<SavedBuckets | string | undefined>(CHOSEN_BUCKET_KEY)
+        return typeof buckets === 'string' ? JSON.parse(buckets) : buckets
```
[the remark this answers is not in the document]
Comment (JadenSimon) [reply src/lambda/wizards/samDeployWizard.ts:204]: `JSON.parse` requires a string type. If we want to guard against anything other than `object | string` then we need to throw an error instead.

**Comment by justinmk3** [comment]
LGTM. Integ CI has failed about 95% of the retries here. Unlikely to be related, but should confirm

**Comment by JadenSimon** [comment]
Looked at each one, they're transient failures. 
One of them should be mitigated by adding a sleep (~10 seconds) before starting the integration tests.
The other is because SAM sometimes rebuilds the image for `target=API` so it ends up building it twice and hitting the timeout time.

**Comment by justinmk3** [comment]
> So for the image + API combo we should skip the build step (or use the `skipPullImage` flag).

Related: https://github.com/aws/aws-toolkit-vscode/issues/376

**Pull request merged**